From: David Teigland <teigland@redhat.com>
Date: Mon, 17 Nov 2008 12:28:48 -0600
Subject: ocfs2: fix wake_up in unlock_ast

In ocfs2_unlock_ast(), call wake_up() on lockres before releasing
the spin lock on it.  As soon as the spin lock is released, the
lockres can be freed.

Signed-off-by: David Teigland <teigland@redhat.com>
Signed-off-by: Mark Fasheh <mfasheh@suse.com>
---
 fs/ocfs2/dlmglue.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

Index: linux-2.6.27-ocfs2/fs/ocfs2/dlmglue.c
===================================================================
--- linux-2.6.27-ocfs2.orig/fs/ocfs2/dlmglue.c
+++ linux-2.6.27-ocfs2/fs/ocfs2/dlmglue.c
@@ -2841,9 +2841,8 @@ static void ocfs2_unlock_ast(void *opaqu
 
 	lockres_clear_flags(lockres, OCFS2_LOCK_BUSY);
 	lockres->l_unlock_action = OCFS2_UNLOCK_INVALID;
-	spin_unlock_irqrestore(&lockres->l_lock, flags);
-
 	wake_up(&lockres->l_event);
+	spin_unlock_irqrestore(&lockres->l_lock, flags);
 
 	mlog_exit_void();
 }
